src/adic_space.lean

```diff
@@ -8,11 +8,132 @@ import for_mathlib.topology
+definition is_hausdorff (α : Type*) [topological_space α] : Prop := 
+  ∀x y, x ≠ y → ∃u v : set α, is_open u ∧ is_open v ∧ x ∈ u ∧ y ∈ v ∧ u ∩ v = ∅
+
+
+open filter lattice
+lemma half_nhd {α : Type*} [topological_space α] [add_group α] [topological_add_group α] (U ∈ (nhds (0 : α)).sets) : 
+  ∃ V ∈ (nhds (0 : α)).sets, ∀ v w ∈ V, v + w ∈ U :=
+/- Here is the story : by continuity of addition, and because 0 + 0 = 0,
+   (+)⁻¹(U) is a neighborhood of (0, 0). Hence it contains some V₁ × V₂. Then set V = V₁ ∩ V₂ -/
+begin
+  have nhdb_in_prod : ((λ a : α × α, a.1+a.2)⁻¹' U) ∈ (nhds ((0, 0) : α × α)).sets,
+    by apply tendsto_add' ; simp [H],
+  rw nhds_prod_eq at nhdb_in_prod,
+  rcases mem_prod_iff.1 nhdb_in_prod with ⟨V₁, H₁, V₂, H₂, H⟩,
+  have H12: V₁ ∩ V₂ ∈ (nhds (0 : α)).sets := inter_mem_sets H₁ H₂,    
+  existsi [(V₁ ∩ V₂), H12],
+  intros v w Hv Hw,
+  have : (v,w) ∈ set.prod V₁ V₂, by finish,
+  simpa using H this
+end
+
+
+instance toplogical_ring.to_uniform_space : uniform_space R := 
+uniform_space.of_core { 
+  uniformity  := (⨅ U ∈ nhd_zero R, principal {p : R×R | p.2 - p.1 ∈ U}),
+  refl        := begin simp, intros i H r, exact mem_of_nhds H end,
+  symm        := tendsto_infi.2 $ assume V, tendsto_infi.2 $ assume h,
+    tendsto_infi' ((λ r, -r) '' V)
+      begin 
+        simp [nhd_zero_symmetric h],
+        unfold tendsto,
+        rw map_principal,
+        have H : prod.swap '' {p : R × R | ∃ (x : R), x ∈ V ∧ -x = p.snd + -p.fst} = {p : R × R | p.snd - p.fst ∈ V},
+        begin 
+          rw set.image_swap_eq_preimage_swap,
+          ext p,
+          cases p with r1 r2,
+          change (∃ x, x ∈ V ∧ -x = r1 - r2) ↔ (r2 - r1 ∈ V),
+          have : ∀ x, -x = r1 - r2 ↔ x = r2 - r1,
+            by intro x; rw [neg_eq_iff_neg_eq, eq_comm]; simp,
+          simp only [this], simp
+      end,
+      simp [H],                           
+    end,
+  comp        := le_infi $ assume V, le_infi $ assume H, 
+    begin
+      rcases half_nhd V H with ⟨U, ⟨h, h'⟩⟩,
+      have comprel : comp_rel {p : R × R | p.snd - p.fst ∈ U} {p : R × R | p.snd - p.fst ∈ U} ⊆ {p : R × R | p.snd - p.fst ∈ V},
+      begin
+        intros p p_comp_rel,
+        rcases p_comp_rel with ⟨z, ⟨Hz1, Hz2⟩⟩,
+        simpa using h' _ _ Hz1 Hz2,
+      end,
+      replace comprel : principal (comp_rel {p : R × R | p.snd - p.fst ∈ U} {p : R × R | p.snd - p.fst ∈ U}) ≤ principal {p : R × R | p.snd - p.fst ∈ V},
+        simpa using comprel,
+      apply lift'_le,
+      swap,
+      exact comprel,
+      apply mem_infi_sets U,
+      apply mem_infi_sets h,
+      simp
+    end }
+
+variable (R)
+/-- Wedhorn Definition 5.31 page 38 -/
+definition is_complete : Prop := is_hausdorff R ∧ ∀ {f:filter R}, cauchy f → ∃x, f ≤ nhds x
+
+variable {R}
+/-- Wedhorn Definition 5.27 page 36 -/
+definition is_bounded 
+  (B : set R) : Prop := ∀ U ∈ nhd_zero R, ∃ V ∈ nhd_zero R, ∀ v ∈ V, ∀ b ∈ B, v*b ∈ U
+
+def power_set (r : R) : set R := set.range (λ n : ℕ, r^n)
```

Comment:
This sounds a lot like "powerset" from basic set theory. Is it an idea to call this merely `powers`? Otherwise, I think it is good to at least include a user comment
```lean
/-- power_set r is the subset of the ring R consisting of all powers of r -/
```

Reply:
Kenny calls them `powers` in his ring theory stuff I think (we localise at the multiplicative set `powers f` IIRC)